pkg/datapath/iptables/custom_chain.go

```diff
@@ -94,87 +97,157 @@ var ciliumChains = []customChain{
-		}
+func (c *customChain) doRename(prog iptablesInterface, newName string) error {
+	if exists, err := c.exists(prog); err != nil {
+		return err
+	} else if !exists {
+		return nil
 	}
-	doRemove := func(c *customChain, prog iptablesInterface, quiet bool) {
-		args := []string{"-t", c.table, "-F", c.name}
-		doProcess(c, prog, args, "flush", quiet)
-		args = []string{"-t", c.table, "-X", c.name}
-		doProcess(c, prog, args, "delete", quiet)
+
+	args := []string{"-t", c.table, "-E", c.name, newName}
+
+	output, err := prog.runProgCombinedOutput(args)
+	if err != nil {
+		return fmt.Errorf("unable to rename %s chain to %s: %s (%w)", c.name, newName, string(output), err)
 	}
+
+	return nil
+}
+
+func (c *customChain) rename(name string) error {
 	if option.Config.EnableIPv4 {
-		doRemove(c, ip4tables, quiet)
+		if err := c.doRename(ip4tables, name); err != nil {
+			return err
+		}
 	}
 	if option.Config.EnableIPv6 && c.ipv6 {
-		doRemove(c, ip6tables, quiet)
+		if err := c.doRename(ip6tables, name); err != nil {
+			return nil
+		}
 	}
+
+	return nil
 }
 
-func getFeedRule(name, args string) []string {
-	ruleTail := []string{"-m", "comment", "--comment", feederDescription + " " + name, "-j", name}
-	if args == "" {
-		return ruleTail
+func (c *customChain) doRemove(prog iptablesInterface) error {
+	if exists, err := c.exists(prog); err != nil {
+		return err
+	} else if !exists {
+		return nil
 	}
-	argsList, err := shellwords.Parse(args)
+
+	args := []string{"-t", c.table, "-F", c.name}
+
+	output, err := prog.runProgCombinedOutput(args)
+	if err != nil {
+		return fmt.Errorf("unable to flush %s chain: %s (%w)", c.name, string(output), err)
+	}
+
+	args = []string{"-t", c.table, "-X", c.name}
+
+	output, err = prog.runProgCombinedOutput(args)
 	if err != nil {
-		log.WithError(err).WithField(logfields.Object, args).Fatal("Unable to parse rule into argument slice")
+		return fmt.Errorf("unable to remove %s chain: %s (%w)", c.name, string(output), err)
 	}
-	return append(argsList, ruleTail...)
+
+	return nil
 }
 
-func (c *customChain) installFeeder() error {
+func (c *customChain) remove() error {
+	if option.Config.EnableIPv4 {
+		if err := c.doRemove(ip4tables); err != nil {
+			return err
+		}
+	}
+	if option.Config.EnableIPv6 && c.ipv6 {
+		if err := c.doRemove(ip6tables); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *customChain) doInstallFeeder(prog iptablesInterface, feedArgs string) error {
 	installMode := "-A"
 	if option.Config.PrependIptablesChains {
 		installMode = "-I"
 	}
 
+	feedRule := []string{"-m", "comment", "--comment", feederDescription + " " + c.name, "-j", c.name}
+	if feedArgs != "" {
+		argsList, err := shellwords.Parse(feedArgs)
+		if err != nil {
+			log.WithError(err).WithField(logfields.Object, feedArgs).Fatal("Unable to parse rule into argument slice")
```

Comment:
Fatal here generally seems out of place. Why does this failure get treated differently? AFAIK, we usually keep Fatal to the daemon subsystem.

Reply:
I do agree that fatal is a bit out of place here, but at the same time I was not sure what the best approach would be as we cannot recover from this error by simply rerunning `doInstallFeeder`, and if we just log it and continue, the iptables ruleset will be most certainly in an inconsistent state (so better fatal out loud and make the error condition clear).

What I can do is just return the error (to remain consistent with the rest of the logic here) and then let the daemon return a fatal error after the 3 attempts